### PR TITLE
Adding folder ID. Considerable refactoring and improvements

### PR DESCRIPTION
**Pull request by philipwhiuk**
~~There's still some stuff outstanding, notably settings migration and a tiny bit of DB migration left to do. But this gets the ball rolling.~~ EDIT: If you can think of missing stuff let me know.

There's I guess 5 parts:

1. Reworking a migration class to use less of the store.
2. Rewriting the POP3 store to match IMAP and WebDav design
3. Testing the heck out of the POP3 store
4. Documentation of the stores in more detail (prep work for EWS)
5. Adding the getID() and using it everywhere it ought to be used.  (Biggest piece of work and responsible for most of the commits).

### CODE REVIEW

**Comment by philipwhiuk** [comment]
Oh and I guarantee there's still a few bugs left to iron out. I need to put it on my device for a good while. Also I want to write more tests for bits and pieces.

**Comment by philipwhiuk** [comment]
DB migration and settings migration is now done. Hopefully all the tests now pass 😄

File: k9mail/src/main/java/com/fsck/k9/preferences/AccountSettings.java [hunk 1]
```diff
@@ -376,4 +379,31 @@ public Integer fromString(String value) throws InvalidSettingValueException {
...
+                    "sentFolder",
+                    "trashFolder",
+                    "spamFolder",
+                    "inboxFolder",
+            };
+
+            for (String setting : settingsToRename) {
+                String value = (String) settings.get(setting + "Name");
+                if (value != null) {
+                    settings.put(setting+"Id", value);
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/preferences/AccountSettings.java]: code style

**Comment by jyn514** [comment]
Do you want me to rebase this real quick? It shouldn't take long to resolve if it's gotten this far.

**Comment by philipwhiuk** [comment]
This needs more than a rebase. It needs a lot of tests to cover various scenarios.

**Comment by philipwhiuk** [comment]
I might try to extract the POP3 rework however - that has considerable value.

**Comment by cketti** [comment]
Superseded by #3236

**Pull request closed without merging**